stats/statsvariables.cpp

```diff
@@ -0,0 +1,1674 @@
+{
+	if (std::find(v.begin(), v.end(), item) == v.end())
+		v.push_back(item);
+}
+
+// Small helper: make a comma separeted list of a vector of QStrings
+static QString join_strings(const std::vector<QString> &v)
+{
+	QString res;
+	for (const QString &s: v) {
+		if (!res.isEmpty())
+			res += ", ";
+		res += s;
+	}
+	return res;
+}
+
+// Note: usually I dislike functions defined inside class/struct
+// declarations ("Java style"). However, for brevity this is done
+// in this rather template-heavy source file more or less consistently.
+
+// Templates to define invalid values for variables and test for said values.
+// This is used by the binners: returning such a value means "ignore this dive".
+template<typename T> T invalid_value();
+template<> int invalid_value<int>()
+{
+	return std::numeric_limits<int>::max();
+}
+template<> double invalid_value<double>()
+{
+	return std::numeric_limits<double>::quiet_NaN();
+}
+template<> QString invalid_value<QString>()
+{
+	return QString();
+}
+template<> StatsQuartiles invalid_value<StatsQuartiles>()
+{
+	double NaN = std::numeric_limits<double>::quiet_NaN();
+	return { NaN, NaN, NaN, NaN, NaN };
+}
+
+static bool is_invalid_value(int i)
+{
+	return i == std::numeric_limits<int>::max();
+}
+
+static bool is_invalid_value(double d)
+{
+	return std::isnan(d);
+}
+
+static bool is_invalid_value(const QString &s)
+{
+	return s.isEmpty();
+}
+
+// Currently, we don't support invalid dates - should we?
+static bool is_invalid_value(const year_quarter &)
+{
+	return false;
+}
+
+static bool is_invalid_value(const dive_site *d)
+{
+	return !d;
+}
+
+static bool is_invalid_value(const StatsOperationResults &res)
+{
+	return !res.isValid();
+}
+
+struct gas_bin_t {
+	enum class Type {
+		Air,
+		Oxygen,
+		EAN,
+		Trimix
+	} type;
+	int o2, he;
+	static gas_bin_t air() {
+		return { Type::Air, 0, 0 };
+	}
+	static gas_bin_t oxygen() {
+		return { Type::Oxygen, 0, 0 };
+	}
+	static gas_bin_t ean(int o2) {
+		return { Type::EAN, o2, 0 };
+	}
+	static gas_bin_t trimix(int o2, int h2) {
+		return { Type::Trimix, o2, h2 };
+	}
+};
+
+static bool is_invalid_value(const gas_bin_t &)
+{
+	return false;
+}
+
+static bool is_invalid_value(const std::vector<StatsValue> &v)
+{
+	return v.empty();
+}
+
+static bool is_invalid_value(const StatsQuartiles &q)
+{
+	return std::isnan(q.min);
+}
+
+bool StatsQuartiles::isValid() const
+{
+	return !is_invalid_value(*this);
+}
+
+// Define an ordering for gas types
+// invalid < air < ean (including oxygen) < trimix
+// The latter two are sorted by (helium, oxygen)
+static bool operator<(const gas_bin_t &t1, const gas_bin_t &t2)
+{
+	if (t1.type != t2.type)
+		return (int)t1.type < (int)t2.type;
+	switch (t1.type) {
+	default:
+	case gas_bin_t::Type::Oxygen:
+	case gas_bin_t::Type::Air:
+		return false;
+	case gas_bin_t::Type::EAN:
+		return t1.o2 < t2.o2;
+	case gas_bin_t::Type::Trimix:
+		return std::tie(t1.o2, t1.he) < std::tie(t2.o2, t2.he);
+	}
+}
```

Comment:
I think it's fine to implement this in two places, but this should mention 
https://github.com/subsurface/subsurface/blob/edd19d4725153e6428e7b2e1b6594b4168601a39/core/divelist.c#L32-L39

Reply:
These are related, but not the same: `get_dive_gas` get the "maximum" gas. The `operator<` is used to establish a total order of gas bins.

Reply:
Right - I was wondering if you want to comment that we use a similar order elsewhere and refer to it. Simply for consistency. I did get that these are different functions :-)

Reply:
Comment added.